File: ElasticPendulumSimulation/tools.py
import random
import time
from math import floor

def show_progress(progress, cc=1, spinner = 0):
    p = floor(progress*20)
    if p <= 8:
        print('[', end='')
        for i in range(p):
            print('#', end='')
        for i in range(8-p):
            print('-', end='')

        print('\033[%dm%03d' % (random.randint(91, 96) if cc == 1 else 93, (progress*100)), end='%\033[0m')
        print('--------]', end='')
    else:
        print('[########', end='')
        print('\033[%dm%03d' % (random.randint(91, 96) if cc == 1 else 93, (progress*100)), end='%\033[0m')
        i=0
        while i in range(p-12):
            print('#', end='')
            i+=1
        while i in range(8):
            print('-', end='')
            i+=1
        print(']', end='')

    if spinner == 1:
        s = int(progress*100) % 9
        if s <= 3:
            print('  /', end = '')
        elif s <= 5:
            print('  -', end = '')
        else:
            print('  \\', end = '')

    if progress == 1:
        print("\r[########100%########]    \033[93mDONE!\033[0m  \n ")
    else:
        print('', end = '\r')
    return

def test(cc):
    for i in range(100+1):
        time.sleep(0.03)
        show_progress(progress=(i/100), cc=cc, spinner=1)




#print('y' if 1>2 else "n")

File: GeneticAlgorithm/cppver/tools.cpp
#include "tools.h"
#include <iomanip>
#include <iostream>
#include <stdio.h>
#include <random>
using namespace std;


list<int> randomchoiceidx(list<float> rate, int k){
    random_device rd;
    mt19937 gen(rd());
    discrete_distribution<> d(rate.begin(), rate.end());
    list<int> choosenidx;// = new list<int>();
    while(choosenidx.size()<k){
        int temp = d(gen);
        bool choosen = 0;
        for(int C : choosenidx){
            if(temp == C){
                choosen = 1;
                break;
            }
        }
        if(choosen == 0){
            choosenidx.push_back(temp);
        }
    }
    return choosenidx;
}



void showprogress(float progress){
    int P = floor(progress*20);
    printf("[");
    for(int i = 1; i <= P; i++){
        printf("#");
    }
    for(int j = 20-P; j > 0; j--){
        printf("-");
    }
    printf("]");
    printf("%6d\r", int(progress*100));
    if(floor(progress)==1){
        printf("[####################]  \033[93mDONE\033[0m\n");
    }
}
